tsl/src/compression/algorithms/uuid_compress.c

```diff
@@ -0,0 +1,625 @@
+		Assert(params.poly[0][0] != 0);
+	}
+
+	uint64 h = umash_full(&params,
+						  /* seed = */ ~0ULL,
+						  /* which = */ 0,
+						  uuid->data,
+						  16);
+
+	return (uint32) (h ^ (h >> 32));
+}
+#else
+static inline uint32
+uuid_compress_hash(pg_uuid_t *uuid)
+{
+	return hash_bytes((unsigned char *) uuid->data, sizeof(*uuid));
+}
+#endif
+
+extern void
+uuid_compressor_append_value(UuidCompressor *compressor, pg_uuid_t next_val)
+{
+	uint64_t components[2];
+	memcpy(components, next_val.data, sizeof(components));
+
+	/* The first component is the timestamp, version and rand_a. */
+	uint64_t timestamp = pg_ntoh64(components[0]);
+	/* The second part is the rand_b and variant. */
+	uint64_t rand_b_and_variant = components[1];
+
+	delta_delta_compressor_append_value(compressor->timestamp, timestamp);
+	uint64_vec_append(&compressor->rand_b_and_variant, rand_b_and_variant);
+
+	uint32 h = uuid_compress_hash(&next_val);
+	addHyperLogLog(&compressor->cardinality, h);
+	compressor->num_values++;
+}
+
+static size_t
+uuid_compressor_estimate_dictionary_storage(UuidCompressor *compressor,
+											size_t nulls_compressed_size)
+{
+	double cardinality = (double) compressor->rand_b_and_variant.num_elements;
+	double cardinality_and_error = cardinality;
+
+	/* Don't use HLL if there are too few elements to estimate the cardinality. */
+	if (cardinality > HLL_MIN_CARDINALITY)
+	{
+		cardinality = estimateHyperLogLog(&compressor->cardinality);
+		cardinality_and_error = cardinality * (1.0 - HLL_ERROR_RATE);
+	}
+
+	int array_index_bytes = ((int) cardinality_and_error * 5 + 63) / 64 * 8;
+
+	double estimated_dictionary_storage =
+		/* 16 bytes per values in dictionary/array/values */
+		cardinality_and_error * 16 +
+		/* a single RLE block for the sizes in dictionary/array/sizes */
+		16 +
+		/* no nulls in dictionary/array/nulls */
+		0 +
+		/* 5 bits on average for the indexes in dictionary/array/indexes */
+		array_index_bytes +
+		/* storing nulls is the same as in the delta-delta compressor */
+		nulls_compressed_size;
+
+	return estimated_dictionary_storage;
+}
+
+extern void *
+uuid_compressor_finish(UuidCompressor *compressor)
+{
+	if (compressor == NULL)
+		return NULL;
+
+	if (compressor->num_values == 0)
+		return NULL;
+
+	size_t nulls_compressed_size = 0;
+	size_t timestamp_compressed_size =
+		delta_delta_compressor_compressed_size(compressor->timestamp, &nulls_compressed_size);
+	size_t estimated_dictionary_storage =
+		uuid_compressor_estimate_dictionary_storage(compressor, nulls_compressed_size);
+	size_t rand_b_and_variant_compressed_size =
+		compressor->rand_b_and_variant.num_elements * sizeof(uint64_t);
+	Assert(compressor->rand_b_and_variant.num_elements == compressor->num_values);
+	size_t total_compressed_size =
+		sizeof(UuidCompressed) + timestamp_compressed_size + rand_b_and_variant_compressed_size;
+
+	/* TODO: this is temporary: to iterate over the delta-delta compressed data
```

Comment:
Sounds very permanent tbh :) Maybe we should store uncompressed data in the compressor first? Allocating extra 16kB once for multiple batches is negligible.